nullaway/src/main/java/com/uber/nullaway/fixserialization/out/ErrorInfo.java

```diff
@@ -22,21 +22,55 @@
 
 package com.uber.nullaway.fixserialization.out;
 
+import com.google.common.collect.ImmutableMap;
 import com.google.errorprone.util.ASTHelpers;
 import com.sun.source.util.TreePath;
 import com.uber.nullaway.ErrorMessage;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 
 /** Stores information regarding an error which will be reported by NullAway. */
 public class ErrorInfo {
 
   private final ErrorMessage errorMessage;
   private final EnclosingClassAndMethodInfo enclosingInfo;
 
+  /** Special characters that need to be escaped in TSV files. */
+  private static final ImmutableMap<Character, Character> escapes =
+      ImmutableMap.of(
+          '\n', 'n',
+          '\t', 't',
+          '\f', 'f',
+          '\b', 'b',
+          '\r', 'r');
+
   public ErrorInfo(TreePath path, ErrorMessage errorMessage) {
     this.enclosingInfo = new EnclosingClassAndMethodInfo(path);
     this.errorMessage = errorMessage;
   }
 
+  /**
+   * Escapes special characters in string to conform with TSV file formats. The most common
+   * convention for lossless conversion is to escape special characters with a backslash according
+   * to <a
+   * href="https://en.wikipedia.org/wiki/Tab-separated_values#Conventions_for_lossless_conversion_to_TSV">
+   * Conventions for lossless conversion to TSV</a>
+   *
+   * @param str String to process.
+   * @return returns modified str which its special characters are escaped.
+   */
+  private static String escapeSpecialCharacters(String str) {
+    // regex needs "\\" to match character '\', each must also be escaped in string to create "\\",
+    // therefore we need four "\".
+    // escape existing backslashes
+    str = str.replaceAll(Pattern.quote("\\"), Matcher.quoteReplacement("\\\\"));
```

Comment:
@nimakarimipour this line means you will replace any occurrence of `\\` with `\\\\`, right?  Why do we need to do this?

Reply:
Ok, I think I understand this and this is correct, though very annoying:

1. The string "\\" in the Java source code here represents in-memory string literal (as `char[]`): [`\`]
2. Similarly, "\\\\" in the source is the literal [`\`,`\`]
3. `Patern.quote("\\")` does the replacement, producing [`\`] -> [`\`,`\`], but this is needed because the first argument of `String.replaceAll(...)` is a regex, which means that literal string matches the string [`\`] (otherwise, a single `\` in the string literal would be interpreted as part of the regex language, waiting for a escaped character inside the regex, leading to an incomplete regex)
4. So far, so good for my understanding of Java regexes. But what I missed the first time around is that, while the second parameter to `String.replaceAll(...)` is **not** a regex, it is something called a "replacement string", with its own special syntax. In particular, the docs have this to say:

> Note that backslashes (\) and dollar signs ($) in the replacement string may cause the results to be different than if it were being treated as a literal replacement string. Dollar signs may be treated as references to captured subsequences as described above, and backslashes are used to escape literal characters in the replacement string.

See [docs](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/regex/Matcher.html#replaceAll(java.lang.String)) for `Matcher.replaceAll`, which is the underlying implementation of `String.replaceAll`.

Thus, for the second argument here: Java source code string `"\\\\"` becomes literal [`\`,`\`], which gets expanded by `Matcher.quoteReplacement` into [`\`,`\`,`\`,`\`], which finally gets interpreted by `String.replaceAll(...)` as a "replacement string" replacing every match by the literal string [`\`,`\`].

Basically this: https://xkcd.com/1638/

But the code looks correct to me :)

Reply:
🤣 great explanation.  Now I'm going to try to forget this stuff exists 🙂

Reply:
@msridhar This actually replaces each occurrence of `\` with `\\` in text. And in order to show `\` in java string, we have to show it with `\\`. The reason for that is that we want the actual `\t` be present in the serialization and not get mixed with the `\t` we inserted ourself. 
<pre>
The output ".equals('\t')     == true" will be serialized to ".equals('\\t')\t == true"
</pre>